### PR TITLE
PLU-318: [TEMPLATES-12]: combine demos and templates (final)

### PR DESCRIPTION
**Pull request by m0nggh**
## Problem
Users are unaware of what they can do with Plumber.

## Solution
Introduce templates to speed up their pipe creation and provide new ideas for them to automate. 

## Details
Combine demos with templates because they are similar.
- Keep demo config except for `isAutoCreated` due to backwards compatibility
- Use tags instead of a tag for each template
- Demo page modal is removed

## Tests
- [x] New user will not have an auto created demo pipe anymore
- [x] Templates are linked properly from empty flows page
- [x] Demo template shows the corresponding demo video upon using template
- [x] Other templates still work as intended without video
- [x] Demos modal cannot be access anymore

### CODE REVIEW

**Comment by m0nggh** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/753?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#753** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/753?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#748** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/748?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 13 other dependent PRs ([#714](https://github.com/opengovsg/plumber/pull/714) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/714?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#715](https://github.com/opengovsg/plumber/pull/715) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/715?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#716](https://github.com/opengovsg/plumber/pull/716) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/716?utm_source=stack-comment-icon" target="_blank">Graphite</a> and 10 others)
* `develop-v2`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @m0nggh and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/backend/src/graphql/custom-resolvers/flow.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  const templateId = parent?.config?.templateConfig.templateId
+  if (!templateId) {
+    return null
+  }
+  return TEMPLATES.find((template) => template.id === templateId)
+}
+
+export default {
+  template,
+} satisfies Resolvers['Flow']
```
Comment (pregnantboy) [inline packages/backend/src/graphql/custom-resolvers/flow.ts]:
```suggestion
} satisfies FlowResolver
```

File: packages/backend/src/helpers/flow-templates.ts [hunk 2]
```diff
@@ -164,36 +162,24 @@ function replaceAllPlaceholdersInValue(
 export async function createFlowFromTemplate(
```
Comment (pregnantboy) [inline packages/backend/src/helpers/flow-templates.ts:162]: refactor as discussed

File: packages/backend/src/models/flow.ts [hunk 3]
```diff
@@ -24,6 +25,7 @@ class Flow extends Base {
   testExecutionId: string
   testExecution?: Execution
   user: User
+  template: Template
```
Comment (pregnantboy) [inline packages/backend/src/models/flow.ts]: remove

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 4]
```diff
@@ -36,33 +35,19 @@ import EditorSnackbar from './EditorSnackbar'
...
-    refetchQueries: [GET_FLOW],
-  })
-
-  const handleClose = useCallback(async () => {
-    await updateFlowConfig()
-  }, [updateFlowConfig])
+  const handleClose = useCallback(() => {
+    searchParams.delete('showDemo') // not using setSearchParams because it creates a browser state
+    window.history.replaceState({}, '', window.location.pathname)
+    window.location.reload()
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx]:
```suggestion
    setSearchParams(searchParams, { replace: true })
```

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 5]
```diff
@@ -124,6 +109,13 @@ export default function EditorLayout() {
     return <InvalidEditorPage />
   }
 
+  // for loading demo video
+  const showDemo = searchParams.get('showDemo')
+  const flowTemplate = flow?.template
+  const isDemoTemplate = flowTemplate?.tags?.some((tag) => tag === 'demo')
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx]: why dont u check for existence of flow.template.demoVideoDetails instead? there's no need to check tags as well right

Comment (pregnantboy) [reply packages/frontend/src/components/EditorLayout/index.tsx]: can move this block after 
if (!flow) check so u dont need the optional chaining

File: packages/frontend/src/components/EditorLayout/index.tsx (hunk 4 above)
Comment (m0nggh) [reply packages/frontend/src/components/EditorLayout/index.tsx]: thanks

File: packages/backend/src/helpers/flow-templates.ts (hunk 2 above)
Comment (m0nggh) [reply packages/backend/src/helpers/flow-templates.ts:162]: thks

**Review by pregnantboy — APPROVED** [review]
lgtm

**Pull request merged**